external_server/server.py

```diff
@@ -0,0 +1,879 @@
+    StatusTimeout,
+    NoPublishedMessage,
+    CommandResponseTimeout,
+    SessionTimeout,
+    UnexpectedMQTTDisconnect,
+)
+from external_server.models.messages import (
+    connect_response as _connect_response,
+    status_response as _status_response,
+)
+from external_server.adapters.mqtt.adapter import MQTTClientAdapter
+from external_server.config import CarConfig as CarConfig, ServerConfig as ServerConfig
+from external_server.models.structures import (
+    GeneralErrorCode,
+    DisconnectTypes,
+    TimeoutType,
+)
+from external_server.models.devices import DevicePy, KnownDevices, device_repr
+from external_server.models.events import (
+    EventType,
+    Event as _Event,
+    EventQueue as _EventQueue,
+)
+from external_server.server_module.server_module import ServerModule as _ServerModule
+from external_server.models.structures import HandledCommand as _HandledCommand
+
+
+eslogger = _ESLogger(LOGGER_NAME)
+carlogger = _CarLogger(LOGGER_NAME)
+
+
+class ServerState(enum.Enum):
+    UNINITIALIZED = enum.auto()
+    CONNECTED = enum.auto()
+    INITIALIZED = enum.auto()
+    RUNNING = enum.auto()
+    STOPPED = enum.auto()
+    ERROR = enum.auto()
+
+
+DeviceStatusName = {
+    _Status.CONNECTING: "CONNECTING",
+    _Status.RUNNING: "RUNNING",
+    _Status.DISCONNECT: "DISCONNECT",
+    _Status.ERROR: "ERROR",
+}
+
+
+StateTransitionTable: dict[ServerState, set[ServerState]] = {
+    ServerState.UNINITIALIZED: {
+        ServerState.ERROR,
+        ServerState.STOPPED,
+        ServerState.CONNECTED,
+    },
+    ServerState.CONNECTED: {
+        ServerState.ERROR,
+        ServerState.STOPPED,
+        ServerState.INITIALIZED,
+    },
+    ServerState.INITIALIZED: {
+        ServerState.ERROR,
+        ServerState.STOPPED,
+        ServerState.RUNNING,
+    },
+    ServerState.RUNNING: {
+        ServerState.ERROR,
+        ServerState.STOPPED,
+    },
+    ServerState.STOPPED: {
+        ServerState.UNINITIALIZED,
+    },
+    ServerState.ERROR: {
+        ServerState.UNINITIALIZED,
+    },
+}
+
+
+ExternalClientMessage = _Connect | _Status | _CommandResponse
+
+
+class ExternalServer:
+
+    def __init__(self, config: ServerConfig) -> None:
+        self._car_servers: dict[str, CarServer] = {}
+        for car_name in config.cars:
+            self._car_servers[car_name] = CarServer(CarConfig.from_server_config(car_name, config))
+        self._company = config.company_name
+
+    @property
+    def company(self) -> str:
+        return self._company
+
+    def car_servers(self) -> dict[str, CarServer]:
+        return self._car_servers.copy()
+
+    def start(self) -> None:
+        threads: dict[str, threading.Thread] = {}
+        for car in self._car_servers:
+            threads[car] = threading.Thread(target=self._car_servers[car].start)
+        for t in threads.values():
+            t.start()
+
+    def stop(self, reason: str = "") -> None:
+        for e in self.car_servers().values():
+            e.stop(reason)
```

Comment:
```suggestion
    def __init__(self, config: ServerConfig) -> None:
        self._car_servers: dict[str, CarServer] = {}
        for car_name in config.cars:
            self._car_servers[car_name] = CarServer(CarConfig.from_server_config(car_name, config))
        self._company = config.company_name
        self._car_threads: list[threading.Thread] = []

    @property
    def company(self) -> str:
        return self._company

    def car_servers(self) -> dict[str, CarServer]:
        return self._car_servers.copy()

    def start(self) -> None:
        for car in self._car_servers:
            self._car_threads.append(threading.Thread(target=self._car_servers[car].start))
        for t in self._car_threads:
            t.start()

    def stop(self, reason: str = "") -> None:
        for e in self.car_servers().values():
            e.stop(reason)
        # Wait for all threads to terminate
        for car_thread in self._car_threads:
            car_thread.join()
        self._car_threads = []
```
Better to wait for threads